packages/berry-pnpify/sources/PnPRootLocator.ts

```diff
@@ -0,0 +1,143 @@
+import * as fs from 'fs';
+
+/**
+ * PnP root locator options
+ */
+export interface PnPLocatorOptions {
+  /**
+   * Function that checks if file exists at given path.
+   *
+   * Default: `fs.existsSync`
+   *
+   * @param filePath file path
+   */
+  existsSync?: (filePath: string) => boolean;
+
+  /**
+   * PnP api filename.
+   *
+   * Default: `.pnp.js`
+   */
+  pnpFileName?: string;
+}
+
+interface DefinedPnPLocatorOptions {
+  existsSync: (filePath: string) => boolean;
+  pnpFileName: string;
+}
+
+/**
+ * A tree structure to minimize the number of fs calls to check for pnp api file existence.
+ *
+ * The nodes of the tree are subdirectory names and the value is either true if the tree path leads
+ * to the PnP API filename or the subtree if the tree path do not lead to PnP API filename.
+ *
+ * Example:
+ *
+ *                             <root>
+ *                               |
+ *                              home
+ *                               |
+ *                              user
+ *                   /                           \
+ *              pnp_project_dir                some_dir
+ *                  |                          /     \
+ *                true        pnp_project_subdir     non_pnp_project_subdir
+ *                                  |                     |          |
+ *                                 true                 dir1       dir2
+ *                                                        |          |
+ *                                                       ...        ...
+ */
+type PnPRootCheckTree = Map<string, any | true>;
+
+/**
+ * PnP root locator given arbitrary path answer the question is this path inside PnP project,
+ * and if yes what is the root directory of this PnP project. If no - it returns null.
+ *
+ * PnP root locator tries to answer this question with minimal possible number of fs calls.
+ *
+ * Assumptions:
+ *  - PnP root cannot be inside `node_modules`
+ *  - PnP root cannot be inside other PnP root
+ */
+export interface PnPRootLocator {
```

Comment:
Why export an interface if a class of the same name is also exported? Documentation?

Reply:
@arcanis It is the requirement of TypeScript compiler. If one removes `export` here, the compiler will output the following error:
`Individual declarations in merged declaration 'PnPRootLocator' must be all exported or all local.ts(2395)`

Reply:
Sure because you declare an interface and a class of the same name, but why have the interface (especially when the class doesn't `implements` it)?